internal/distro/rhel9/imagetype.go

```diff
@@ -273,13 +287,62 @@ func (t *imageType) PartitionType() string {
 	return basePartitionTable.Type
 }
 
+func (t *imageType) initializeManifest(bp *blueprint.Blueprint,
+	options distro.ImageOptions,
+	repos []rpmmd.RepoConfig,
+	packageSets map[string]rpmmd.PackageSet,
+	containers []container.Spec,
+	seed int64) (*manifest.Manifest, error) {
+
+	if err := t.checkOptions(bp.Customizations, options, containers); err != nil {
+		return nil, err
+	}
+
+	// TODO: let image types specify valid workloads, rather than
+	// always assume Custom.
+	w := &workload.Custom{
+		BaseWorkload: workload.BaseWorkload{
+			Repos: packageSets[blueprintPkgsKey].Repositories,
+		},
+		Packages: bp.GetPackagesEx(false),
+	}
```

Comment:
Note: We will actually have to implement a way to merge workloads. The reason is that if we define SAP as a workload, instead of an image type, then we would still have to have a way for user to customize image with SAP workload with potentially additional packages, enabled services, etc... Also the complication is that we now depsolve packages in chain where the workload packages are always depsolved as a 2nd transaction, which may not be desirable for "standard" workloads like SAP, but really only for packages provided by the user in BP. Probably not that relevant for your rework, but this will have to be figured out before we use Workloads for anything else than BP customizations.

Reply:
My thinking with workloads was that we would not expose the same options in SAP as Custom, but only those options that make sense in that setting. I imagine that for an SAP image we would not support adding extra packages, systems services etc, maybe we would not support any options at all, or just a few SAP specific ones. The benefit of the named workloads would be that we know the use case so can be more restrictive.

Reply:
hmm, I see. It is definitely an option. If that would be the case, then things could stay as they are. However it would also mean regressing in our current functionality, because we support customization of any image type, even those that are focused on specific workload.

So as a result, I expected that we want to allow building custom images of any type. I can imagine that the user may need to install some additional packages and services also into predefined-workload-specific images for monitoring, inspection of the system, etc. E.g. also we do forward logs to Splunk, the customer's security team may require a special software to be installed on the OS for CVE and threat assessment.

With the SAP image as an example, it is preconfigured for SAP, but does not come with SAP software. Maybe someone would want to build custom SAP images with the actual SAP software installed from custom repos 🤔 I'm not actually sure, just describing situations in which it would make sense to allow (package and service) customizations. The workload-specific customizations are definitely an interesting idea and IMO would make sense.

Reply:
I agree, I also don't know for certain. It could be that we'll need a lot (all?) the customisations for some specific workloads, but I think we should have the discussion on a case-by-case basis on what makes sense, to limit our test exposure etc.

Do we already expose SAP images with customisations anywhere? I thought they were only built internally?

Reply:
> Do we already expose SAP images with customisations anywhere? I thought they were only built internally?

The image type is filtered out by default via Weldr API. With a configuration change and adding a few new repo definitions for SAP-specific packages would definitely make it possible to build it on-premise. Nevertheless this is not really documented anywhere and would require knowing what to do. So we can IMO safely consider the SAP image in general not being buildable on-premise.

For Clous API (including the internal service), any customizations can be already passed when building the SAP image (as well as any other image). AFAIK it is not really used by anyone nor by any pipeline. My point was that it is possible, not that it is used.

Reply:
Thanks for the context!